pkg/queryfrontend/roundtrip.go

```diff
@@ -237,6 +235,24 @@ func newQueryRangeTripperware(
 	}, nil
 }
 
+func dynamicIntervalFn(config QueryRangeConfig) queryrange.IntervalFn {
+	return func(r queryrange.Request) time.Duration {
+		// Use static interval, by default.
+		if config.SplitQueriesByInterval != 0 {
+			return config.SplitQueriesByInterval
+		}
+
+		queryInterval := time.Duration(r.GetEnd()-r.GetStart()) * time.Millisecond
+		// if the query is multiple of max interval, we use the max interval to split.
+		if queryInterval/config.MaxQuerySplitInterval >= 2 {
+			return config.MaxQuerySplitInterval
+		}
+
+		// if the query duration is less than max interval, we split it equally in HorizontalShards.
+		return time.Duration(queryInterval.Milliseconds()/config.HorizontalShards) * time.Millisecond
```

Comment:
I don't see where `MinQuerySplitInterval` is used here. We need to also ensure the query interval is larger than `MaxQuerySplitInterval` IIUC?

Reply:
`MinQuerySplitInterval` is a drop-in replacement for `SplitQueriesByInterval`, saying just when we start splitting queries. We use it in `newQueryRangeTripperware` (line 182) to decide whether to send split requests or to send the original request.

Regarding the `MaxQuerySplitInterval`, we are checking that on line 247 to see if the query interval is at least twice the size of the `MaxQuerySplitInterval`, if so we use that as our split interval.

Reply:
> MinQuerySplitInterval is a drop-in replacement for SplitQueriesByInterval, saying just when we start splitting queries. We use it in newQueryRangeTripperware (line 182) to decide whether to send split requests or to send the original request.

Where do we ensure that the query range is larger than the `MinQuerySplitInterval`? For this code, I can see that even a small range query will be sharded by `HorizontalShards` times.

Reply:
Yeah, you are right. I added new test case and an if statement.